src/D2L.CodeStyle.Analysis/MutabilityInspector.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+
+namespace D2L.CodeStyle.Analysis {
+
+	public sealed class MutabilityInspector {
+
+		/// <summary>
+		/// A list of known non-valuetype immutable types
+		/// </summary>
+		private static readonly ImmutableHashSet<string> KnownImmutableTypes = new HashSet<string> {
+			nameof(String)
+		}.ToImmutableHashSet();
+
+		/// <summary>
+		/// Determine if a given type is immutable.
+		/// </summary>
+		/// <param name="type">The type to determine immutability for.</param>
+		/// <returns>Whether the type is immutable.</returns>
+		public bool IsTypeMutable(
+			ITypeSymbol type
+		) {
+			if( type.IsValueType ) {
+				return false;
+			}
+
+			if( type.TypeKind == TypeKind.Array ) {
+				return true;
+			}
+
+			if( KnownImmutableTypes.Contains( type.Name ) ) {
+				return false;
+			}
+
+			foreach( var member in type.GetMembers() ) {
+				if( member is IPropertySymbol && IsPropertyMutable( member as IPropertySymbol ) ) {
+					return true;
+				}
+				if( member is IFieldSymbol && IsFieldMutable( member as IFieldSymbol ) ) {
+					return true;
+				}
+				// skip methods, events, ctors, etc.
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		/// Determine if a property is externally immutable (private properties are considered immutable).
+		/// This does not check if the type of the property is also immutable; use <see cref="MutabilityInspector.IsTypeMutable"/> for that.
+		/// </summary>
+		/// <param name="prop">The property to check for mutability.</param>
+		/// <returns>Determines whether the property is immutable.</returns>
+		public bool IsPropertyMutable( IPropertySymbol prop ) {
+			if( prop.DeclaredAccessibility == Accessibility.Private ) {
+				return false;
+			}
+			if( prop.IsReadOnly ) {
+				return false;
+			}
+			if( prop.SetMethod.DeclaredAccessibility == Accessibility.Private ) {
+				return false;
+			}
+			return true;
+		}
+
+		/// <summary>
+		/// Determine if a field is externally immutable (private fields are considered immutable).

```

Comment:
I think you're right. I wanted to separate the public and private interfaces because we will have problems with interface types (where we can only inspect the public interface), but I agree that opens a particular class of problems. 

Interfaces can give us false positives like this:

```csharp
public class ClientNameFetcher {
    private readonly IDbFactory m_dbFactory;

    public ClientNameFetcher( IPluginsManger plugins ) {
         // as an example of something that is expensive, so this class should be cached, either as a singleton or as a static -- regardless its lifetime should be longer than the thing using it, and there is possibility of data leakage in general; remember that singletons and statics behave the same as far as leakage is concerned
        m_dbFactory = plugins.GetPlugin<IDbFactory>().First(); 
    }
    public string GetNameOfClient( Guid clientGuid ) {
        return m_dbFactory.GetDb( clientGuid, "Main" ).GetClientName();
    }
}
public class ClientNameController {
    private static readonly ClientNameFetcher = new ClientNameFetcher(); // or injected in, where its registered as a singleton
    public string Get( Guid clientGuid ) {
        return m_clientNameFetcher.Get( clientGuid );
    }
}
```
In this example (which is a problem if `ClientNameFetcher.m_dbFactory` holds client-specific data internally), we can't verify that `ClientNameController.ClientNameFetcher` is immutable, because the type can't be verified to be immutable, because `ClientNameFetcher.m_dbFactory` is an interface, and we can only reason about its public interface, not its private implementation. I wanted to look only at the public interface for this reason, but you're right, we can be smarter about private data where we have it. 

I'll fix this.